pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,212 @@
+	if a.keyType != RSA {
+		return 0, fmt.Errorf("unable to retrieve RSA key size for key type: %T", a.keyType)
+	}
+	rsaKeySize, ok := a.extraKeyParams.(RSAKeySize)
+	if !ok {
+		panic("unable to retrieve key size for RSA, malformed algorithm details?")
+	}
+	return rsaKeySize, nil
+}
+
+func (a algorithmDetailsImpl) GetECDSACurve() (*elliptic.Curve, error) {
+	if a.keyType != ECDSA {
+		return nil, fmt.Errorf("unable to retrieve ECDSA curve for key type: %T", a.keyType)
+	}
+	ecdsaCurve, ok := a.extraKeyParams.(elliptic.Curve)
+	if !ok {
+		panic("unable to retrieve curve for ECDSA, malformed algorithm details?")
+	}
+	return &ecdsaCurve, nil
+}
+
+func (a algorithmDetailsImpl) checkKey(pubKey crypto.PublicKey) bool {
+	switch a.keyType {
+	case RSA:
+		rsaKey, ok := pubKey.(rsa.PublicKey)
+		if !ok {
+			return false
+		}
+		keySize, err := a.GetRSAKeySize()
+		if err != nil {
+			panic(err)
+		}
+		return rsaKey.Size() == int(keySize)
+	case ECDSA:
+		ecdsaKey, ok := pubKey.(ecdsa.PublicKey)
+		if !ok {
+			return false
+		}
+		curve, err := a.GetECDSACurve()
+		if err != nil {
+			panic(err)
+		}
+		return ecdsaKey.Curve == *curve
+	case ED25519:
+		_, ok := pubKey.(ed25519.PublicKey)
+		return ok
+	}
+	panic("unreachable")
+}
+
+func (a algorithmDetailsImpl) checkHash(hashType crypto.Hash) bool {
+	return a.hashType == hashType
+}
+
+var algorithmDetails = []algorithmDetailsImpl{
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_2048_SHA256, RSA, crypto.SHA256, RSAKeySize(2048), "rsa-sign-pkcs1-2048-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_3072_SHA256, RSA, crypto.SHA256, RSAKeySize(3072), "rsa-sign-pkcs1-3072-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_4096_SHA256, RSA, crypto.SHA256, RSAKeySize(4096), "rsa-sign-pkcs1-4096-sha256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_256_NISTP256, ECDSA, crypto.SHA256, elliptic.P256(), "ecdsa-sha2-256-nistp256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_384_NISTP384, ECDSA, crypto.SHA384, elliptic.P384(), "ecdsa-sha2-384-nistp384"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_512_NISTP521, ECDSA, crypto.SHA512, elliptic.P521(), "ecdsa-sha2-512-nistp521"},
+	{v1.KnownSignatureAlgorithm_ED25519, ED25519, crypto.Hash(0), nil, "ed25519"},
+	{v1.KnownSignatureAlgorithm_ED25519_PH, ED25519, crypto.Hash(0), nil, "ed25519-ph"},
+}
+
+// AlgorithmRegistryConfig represents a set of permitted algorithms for a given Sigstore service or component.
+//
+// Individual services may wish to restrict what algorithms are allowed to a subset of what is covered in the algorithm
+// registry (represented by v1.KnownSignatureAlgorithm).
+type AlgorithmRegistryConfig struct {
+	permittedAlgorithms []AlgorithmDetails
+}
+
+// GetAlgorithmDetails retrieves a set of details for a given v1.KnownSignatureAlgorithm flag that allows users to
+// introspect the public key algorithm, hash algorithm and more.
+func GetAlgorithmDetails(knownSignatureAlgorithm v1.KnownSignatureAlgorithm) (AlgorithmDetails, error) {
```

Comment:
Does this need to be public?

Reply:
I think not, assuming you don't need it 🙂